Lowrisc zeroisation top level dv

antmarzam opened the pull request:
Adding top-level tests for zeroization. 
This PR covers tests from issue: https://github.com/chipsalliance/caliptra-ss/issues/603
1 - first commit
2 - second commit
5 - TODO
6 - TODO

andreaskurth commented on src/integration/test_suites/smoke_test_fc_ratchet_seed_write_read/smoke_test_fc_ratchet_seed_write_read.c:
```diff
@@ -0,0 +1,404 @@
+    for (uint32_t addr = partitions[CPTRA_SS_LOCK_HEK_PROD_0].address; addr < partitions[CPTRA_SS_LOCK_HEK_PROD_0].digest_address; addr += 4) {
+      dai_wr(addr, w_data[0], w_data[1], 32, 0);
+      dai_rd(addr, &r_data[0], &r_data[1], 32, 0);
+      if (r_data[0] != w_data[0] && r_data[1] != w_data[1] ) {
```
Doesn't this lead to an error? I'm surprised because as you `dai_wr()` with `granularity=32`, the second 32 bit word (`w_data[1]`) does not get written.

andreaskurth commented on src/integration/test_suites/smoke_test_fc_ratchet_seed_write_read/smoke_test_fc_ratchet_seed_write_read.c:
```diff
@@ -0,0 +1,404 @@
+    //=================================
+    // Calculate digests:
+    //=================================
+    dai_wr(partitions[CPTRA_SS_LOCK_HEK_PROD_0].digest_address, 0xFF, 0xFF, 32, 0);
```
Digests have a granularity of 64 bit:
```suggestion
    dai_wr(partitions[CPTRA_SS_LOCK_HEK_PROD_0].digest_address, 0xFF, 0xFF, 64, 0);
```

antmarzam replied:
Done

andreaskurth commented on src/integration/test_suites/smoke_test_fc_ratchet_seed_write_read/smoke_test_fc_ratchet_seed_write_read.c:
```diff
@@ -0,0 +1,404 @@
+    // Zeroize marker field.
+    dai_zer(partitions[CPTRA_SS_LOCK_HEK_PROD_0].zer_address, &r_data[0], &r_data[1], 64, 0);
+    if (r_data[0] != 0xFFFFFFFF || r_data[1] != 0xFFFFFFFF) {
+      VPRINTF(LOW, "ERROR: [partitions[CPTRA_SS_LOCK_HEK_PROD_0]  digest is not zeroized (post-digest)\n");
```
This is the zeroization marker field, not the digest. The digest can (and should) also be zeroized in this test

antmarzam replied:
I've cleaned the printouts too in my latest push. The digest is now being zeroized

andreaskurth commented on src/integration/test_suites/smoke_test_fc_ratchet_seed_volatile_lock/smoke_test_fc_ratchet_seed_volatile_lock.c:
```diff
@@ -0,0 +1,187 @@
+    VPRINTF(LOW, "==================================\n testing all ratchet seed registers: \n")
+    VPRINTF(LOW, "testing all ratchet seed registers. ratcher_seed_volatile_lock = 0 \n==================================\n\n")
+    // LOCK set to 0x0 - all partitions are writable
+    dai_wr(partitions[CPTRA_SS_LOCK_HEK_PROD_0].address, 0xFF, 0xFF, 32, 0);
```
We should do a `dai_rd` to check what was written. Please note that due to the 32-bit granularity of this partition, only the first word will be written.

andreaskurth commented on src/integration/test_suites/smoke_test_fc_secret_partition_zeroization/smoke_test_fc_secret_partition_zeroization.c:
```diff
@@ -0,0 +1,160 @@
+ * partition and verify that the valid bit is asserted once the partition has been locked.
+ * The test proceeds in the following steps:
+ *
+ *   1. Write a full transition token into the partition.
+ *   2. Read back the token and verify it is equal to the token written in Step 1.
+ *      This works as long as the secret partition has not been locked.
+ *   3. Verify that the partition's digest register is 0.
+ *   4. Lock the partition by calculating a hardware digest.
+ *   5. Reset the RTL.
+ *   6. Try to read the token and verify that it results in an error as the
+ *      partition is secret and locked, thus not accessible by software anymore.
+ *   7. Try to write a value into the partition and verify that it results in an error
+ *      as the partition has been locked in Step 5.
+ *   8. Read back the digest from the partition's digest register and verify it
+ *      is non-zero.
+ *   9. Apply zeroization
+ *   10.Verify the valid signal drops via SVA.
```
Nice! I suggest placing those comments in the code below, as in
```c
// Step 1: Write a full transition token into the partition.
```
as this will be less likely to become outdated and is easier to find when reading the code

antmarzam replied:
Thanks - all credit goes to @andrea-caforio since I've based what I've written from his top-level tests

antmarzam replied:
Please have a look at the latest push and see what you think

andreaskurth commented on src/integration/test_suites/smoke_test_fc_secret_partition_zeroization/smoke_test_fc_secret_partition_zeroization.c:
```diff
@@ -0,0 +1,160 @@
+    dai_rd(fuse_address, &read_data[0], &read_data[1], 64, 0);
+    dai_rd(fuse_address+8, &read_data[2], &read_data[3], 64, 0);
+    if (memcmp(data, read_data, 16)) {
+        VPRINTF(LOW, "ERROR: incorrect fuse data: expected: %08X actual: %08X\n", data[2], read_data[2]);
```
This would have to be a loop over all four 32-bit word pairs to be a complete error message

(Not super important though)

antmarzam replied:
Comparing the 128-bit - Fixed in the latest push

andreaskurth commented on src/integration/test_suites/smoke_test_fc_secret_partition_zeroization/smoke_test_fc_secret_partition_zeroization.c:
```diff
@@ -0,0 +1,160 @@
+    // Step 8
+    digest[0] = lsu_read_32(SOC_OTP_CTRL_SECRET_LC_TRANSITION_PARTITION_DIGEST_DIGEST_0);
+    digest[1] = lsu_read_32(SOC_OTP_CTRL_SECRET_LC_TRANSITION_PARTITION_DIGEST_DIGEST_1);
+    if (digest[0] == 0 && digest[1] == 0) {
```
This needs to be an `||`

andreaskurth commented on src/integration/test_suites/smoke_test_fc_secret_partition_zeroization/smoke_test_fc_secret_partition_zeroization.c:
```diff
@@ -0,0 +1,160 @@
+        exit(1);
+    }
+
+    // Step 9 : Apply zeroization
```
We should also zeroize the digest

andreaskurth reviewed (changes requested):
Thx @antmarzam. I think this goes in the right direction, although a few changes are needed

engdoreis commented on src/integration/test_suites/smoke_test_fc_ratchet_seed_write_read/smoke_test_fc_ratchet_seed_write_read.c:
```diff
@@ -0,0 +1,404 @@
+
+void ratchet_seed_write_read(void) {
+    uint32_t r_data[2];
+    uint32_t w_data[2];
+
+    w_data[0] = 0x3;
+    w_data[1] = 0x3;
```
```suggestion
    uint32_t w_data[2] = {0x03, 0x03};
```

antmarzam replied:
Thanks - this should be present in my latest push

ekarabu commented on src/integration/test_suites/smoke_test_fc_secret_partition_zeroization/smoke_test_fc_secret_partition_zeroization.c:
```diff
@@ -0,0 +1,160 @@
+#endif
+
+/**
+ * Program a LCC transition token in `SECRET_LC_TRANSITION_PARTITION` zeroize the secret
```
As we agreed earlier, the only zeroizable secret partition is vendor secret partition.

andreaskurth commented on src/integration/test_suites/smoke_test_fc_ratchet_seed_write_read/smoke_test_fc_ratchet_seed_write_read.c:
```diff
@@ -0,0 +1,173 @@
+uint32_t write_read_entire_partition(uint32_t num_partition, uint32_t wdata0, uint32_t wdata1, uint32_t write_error_mask, uint32_t read_error_mask) {
+  uint32_t rdata[2];
+
+  for (uint32_t addr = partitions[num_partition].address; addr < partitions[num_partition].digest_address; addr += 4) {
```
For generality, since you're (correctly) using `granularity` below:
```suggestion
  for (uint32_t addr = partitions[num_partition].address; addr < partitions[num_partition].digest_address; addr += partitions[num_partition].granularity / 8) {
```

antmarzam replied:
Done

andreaskurth commented on src/integration/test_suites/smoke_test_fc_ratchet_seed_write_read/smoke_test_fc_ratchet_seed_write_read.c:
```diff
@@ -0,0 +1,173 @@
+
+  memset(r_data, 0, 2*sizeof(uint32_t));
+  // Zeroize fuse.
+  for (uint32_t addr = partitions[num_partition].address; addr < partitions[num_partition].digest_address; addr += 4) {
```
```suggestion
  for (uint32_t addr = partitions[num_partition].address; addr < partitions[num_partition].digest_address; addr += partitions[num_partition].granularity / 8) {
```

ekarabu commented on src/integration/test_suites/smoke_test_fc_ratchet_seed_volatile_lock/smoke_test_fc_ratchet_seed_volatile_lock.c:
```diff
@@ -0,0 +1,132 @@
+void volatile_lock_set_write_read_ratchet(uint32_t lock_value, uint32_t wdata, uint32_t partition_locked) {
+  uint32_t mask, rdata;
+
+  lsu_write_32(SOC_OTP_CTRL_RATCHET_SEED_VOLATILE_LOCK, lock_value);
```
what is this lock value? Do we have just one lock register for all ratchet seeds?

antmarzam commented on src/integration/test_suites/smoke_test_fc_ratchet_seed_write_read/smoke_test_fc_ratchet_seed_write_read.c:
```diff
@@ -0,0 +1,173 @@
+  // Zeroize fuse.
+  for (uint32_t addr = partitions[num_partition].address; addr < partitions[num_partition].digest_address; addr += 4) {
+    dai_zer(addr, &r_data[0], &r_data[1], partitions[num_partition].granularity, 0);
+    if (r_data[0] != 0xFFFFFFFF) {
```
Done

antmarzam commented on src/integration/test_suites/smoke_test_fc_ratchet_seed_write_read/smoke_test_fc_ratchet_seed_write_read.c:
```diff
@@ -0,0 +1,173 @@
+/**
+   This tests:
+   1 - writes to each ratchet seed one by one and reads back the value written was returned.
+   2 - calculates the digest for all HEKs
```
Done (also added the part of reading the digest)

antmarzam commented on src/integration/test_suites/smoke_test_fc_ratchet_seed_write_read/smoke_test_fc_ratchet_seed_write_read.c:
```diff
@@ -0,0 +1,173 @@
+    // Step 2 - calculates the digest for all HEKs - which locks the partition
+    for (uint32_t partition_no = CPTRA_SS_LOCK_HEK_PROD_0; partition_no <= CPTRA_SS_LOCK_HEK_PROD_7; partition_no++) {
+      dai_wr(partitions[partition_no].digest_address, 0xFF, 0xFF, 32, 0);
+    }
```
I've added the reads after writing the digests

antmarzam commented on src/integration/test_suites/smoke_test_fc_ratchet_seed_write_read/smoke_test_fc_ratchet_seed_write_read.c:
```diff
@@ -0,0 +1,173 @@
+      dai_wr(partitions[partition_no].digest_address, 0xFF, 0xFF, 32, 0);
+    }
+    VPRINTF(LOW, "==========================================\n");
+    VPRINTF(LOW, "All the digests for HEKs partitions have been calculated\n");
```
Done

antmarzam commented on src/integration/test_suites/smoke_test_fc_ratchet_seed_volatile_lock/smoke_test_fc_ratchet_seed_volatile_lock.c:
```diff
@@ -0,0 +1,132 @@
+/**
+   This tests takes the following steps:
+   1 - Write read each ratchet seed while testing all possible values
+   2 - calculates the digest, and resets the device
```
Done

antmarzam commented on src/integration/test_suites/smoke_test_fc_secret_partition_zeroization/smoke_test_fc_secret_partition_zeroization.c:
```diff
@@ -0,0 +1,167 @@
+    // Step 9. Apply zeroization
+
+    // Zeroize fuse.
+    for (uint32_t addr = partitions[VENDOR_SECRET_PROD_PARTITION].address; addr < partitions[VENDOR_SECRET_PROD_PARTITION].digest_address; addr += 4) {
```
Done

antmarzam commented on src/integration/test_suites/smoke_test_fc_secret_partition_zeroization/smoke_test_fc_secret_partition_zeroization.c:
```diff
@@ -0,0 +1,167 @@
+      goto epilogue;
+    }
+    memset(read_data, 0, 2*sizeof(uint32_t));
+
```
I've added a new variable `zeroize_num_times` I've set it to 3 for now, and I've wrapped the zeroization loop around it

antmarzam commented on src/integration/test_suites/smoke_test_fc_partition_write_zeroize/smoke_test_fc_partition_write_zeroize.c:
```diff
@@ -0,0 +1,118 @@
+        goto epilogue;
+      }
+    }
+
```
digest and reset are here: https://github.com/chipsalliance/caliptra-ss/blob/5afadd0ebb2ae91e0cc0bee578aed722f3b0df3a/src/integration/test_suites/smoke_test_fc_secret_partition_zeroization/smoke_test_fc_secret_partition_zeroization.c#L93-L98

The zeroization happens at the end in step#9

antmarzam replied:
I've push changes adding digest and reset before the zeroization. 
It would make the test take longer and that's already tested elsewhere. Do you want the changes in this test anyway?

engdoreis commented on src/integration/test_suites/smoke_test_fc_ratchet_seed_write_read/smoke_test_fc_ratchet_seed_write_read.c:
```diff
@@ -0,0 +1,180 @@
+    dai_rd(addr, &rdata[0], &rdata[1], partitions[num_partition].granularity, read_error_mask);
+    if (rdata[0] != wdata0 || (partitions[num_partition].granularity > 32 && rdata[1] != wdata1)) {
+      VPRINTF(LOW, "ERROR: partitions[num_partition=%0d] read_value, does not match written_value\n", num_partition);
+      return 0;
```
Nit, I would recomend changing the return type to bool or return 0 for success and -1 for error.

engdoreis commented on src/integration/test_suites/smoke_test_fc_ratchet_seed_write_read/smoke_test_fc_ratchet_seed_write_read.c:
```diff
@@ -0,0 +1,180 @@
+  uint32_t r_data[2];
+
+
+  memset(r_data, 0, 2*sizeof(uint32_t));
```
```suggestion
  memset(r_data, 0, sizeof(r_data));
  ```
  The same for the other `memset` bellow.

engdoreis commented on src/integration/test_suites/smoke_test_fc_ratchet_seed_write_read/smoke_test_fc_ratchet_seed_write_read.c:
```diff
@@ -0,0 +1,180 @@
+
+    // Step 3 - tries writing and reading and checking a DAI error is returned
+    for (uint32_t partition_no = CPTRA_SS_LOCK_HEK_PROD_0; partition_no <= CPTRA_SS_LOCK_HEK_PROD_7; partition_no++) {
+      if (!write_read_entire_partition(partition_no, w_data[0], w_data[1], OTP_CTRL_STATUS_DAI_ERROR_MASK, 0))
```
I suggest avoiding code blocks without braces `{}`, the same for line 153.

antmarzam replied:
I've added the curly braces as per your suggestion - thanks!

engdoreis commented on src/integration/test_suites/smoke_test_fc_ratchet_seed_volatile_lock/smoke_test_fc_ratchet_seed_volatile_lock.c:
```diff
@@ -0,0 +1,141 @@
+  for (uint32_t partition_no = CPTRA_SS_LOCK_HEK_PROD_0; partition_no <= CPTRA_SS_LOCK_HEK_PROD_7; partition_no++) {
+    mask = (((partition_no-partition_index_offset) < lock_value) ||  partition_locked) ? OTP_CTRL_STATUS_DAI_ERROR_MASK : 0;
+
+      dai_wr(partitions[partition_no].address, wdata0, wdata1 , partitions[partition_no].granularity, mask);
```
nit. the identation looks wrong here.

engdoreis commented on src/integration/test_suites/smoke_test_fc_secret_partition_zeroization/smoke_test_fc_secret_partition_zeroization.c:
```diff
@@ -0,0 +1,170 @@
+  for (uint32_t idx = 0; idx < 4; idx++) {
+    if (data[idx] != read_data[idx]) {
+      VPRINTF(LOW, "ERROR: incorrect fuse data[index=%0d]: expected: %08X actual: %08X\n", idx, data[idx], read_data[idx]);
+      exit(1);
```
why `exit` and not `goto epilogue`?

engdoreis commented on src/integration/test_suites/smoke_test_fc_secret_partition_zeroization/smoke_test_fc_secret_partition_zeroization.c:
```diff
@@ -0,0 +1,170 @@
+          VPRINTF(LOW, "ERROR: [partitions[SECRET_LC_TRANSITION_PARTITION] ]fuse is not zeroized\n");
+          goto epilogue;
+        }
+        }
```
The identation nees fix.

antmarzam replied:
Fixed

engdoreis commented on src/integration/test_suites/smoke_test_fc_secret_partition_zeroization/smoke_test_fc_secret_partition_zeroization.c:
```diff
@@ -0,0 +1,170 @@
+          goto epilogue;
+        }
+        }
+      memset(read_data, 0, 2*sizeof(uint32_t));
```
Please see my suggestion to memset on previous commit.

engdoreis commented on src/integration/test_suites/smoke_test_fc_partition_write_zeroize/smoke_test_fc_partition_write_zeroize.c:
```diff
@@ -0,0 +1,141 @@
+
+    // Step 2 - Write, then read the digest for HEK0 - which locks the partition
+    dai_wr(partitions[CPTRA_SS_LOCK_HEK_PROD_0].digest_address, 0xFF, 0xFF, 64, 0);
+    memset(r_data, 0, 2*sizeof(uint32_t));
```
Please see my suggestion to memset on previous commit.

engdoreis reviewed (commented):
Thanks for the many changes @antmarzam, now it'ss much easier to read the test.
I left a few suggestion.

antmarzam commented:
@andreaskurth / @andreaskurth  / @ekarabu : Can you please have another look to see if you're happy with the changes? I believe I've now addressed all your comments. Thanks!

moidx commented on src/integration/test_suites/smoke_test_fc_secret_partition_zeroization/smoke_test_fc_secret_partition_zeroization.c:
```diff
@@ -0,0 +1,168 @@
+  }
+  // Step 3. Verify that the partition's digest register is 0.
+  uint32_t digest[2];
+  digest[0] = lsu_read_32(OTP_CTRL_SECRET_PROD_PARTITION_0_DIGEST_DIGEST_0);
```
This reads will fail because the offsets are not relative to the SOC memory map used by the test harness. 

Perhaps should read from `SOC_OTP_CTRL_SECRET_PROD_PARTITION_0_DIGEST_DIGEST_0` instead?

moidx replied:
I thought the digest registers were populated after reset. Is it possible to actually read them before reset?

moidx commented on src/integration/test_suites/smoke_test_fc_secret_partition_zeroization/smoke_test_fc_secret_partition_zeroization.c:
```diff
@@ -0,0 +1,168 @@
+
+void secret_partition_zeroization(void) {
+  const uint32_t base_address = partitions[VENDOR_SECRET_PROD_PARTITION].address;
+  const uint32_t fuse_address = CPTRA_SS_TEST_EXIT_TO_MANUF_TOKEN;
```
`CPTRA_SS_TEST_EXIT_TO_MANUF_TOKEN` seems to be in the `SECRET_LC_TRANSITION_PARTITION`. This means that the digest operation below is being applied to the wrong partition. Can you check this?

moidx commented on src/integration/test_suites/smoke_test_fc_secret_partition_zeroization/smoke_test_fc_secret_partition_zeroization.c:
```diff
@@ -0,0 +1,168 @@
+
+
+void secret_partition_zeroization(void) {
+  const uint32_t base_address = partitions[VENDOR_SECRET_PROD_PARTITION].address;
```
Should this be `SECRET_LC_TRANSITION_PARTITION`?

moidx replied:
Actually, since this partition is non-zeroizable we'll have to update the test implementation to use one of the correct fields - e.g. `CPTRA_SS_VENDOR_SPECIFIC_SECRET_FUSE_0`. You'll have to check the hjson overrides as well.

moidx commented on src/integration/test_suites/smoke_test_fc_secret_partition_zeroization/smoke_test_fc_secret_partition_zeroization.c:
```diff
@@ -0,0 +1,168 @@
+  uint32_t zeroize_num_times = 3;
+
+  // Step 1. Write a full transition token into the partition.
+  dai_wr(fuse_address, data[0], data[1], 64, 0);
```
The default OTP image seems to be intialized with random data at the `CPTRA_SS_TEST_EXIT_TO_MANUF_TOKEN` offset. You may have to create a separate image with the values left uninitialized to be able to run this test without ECC errors.

moidx commented on src/integration/test_suites/smoke_test_fc_secret_partition_zeroization/smoke_test_fc_secret_partition_zeroization.c:
```diff
@@ -0,0 +1,168 @@
+
+  // Step 7. Try to write a value into the partition and verify that it results in an error
+  //         as the partition has been locked in Step 5.
+  dai_wr(fuse_address, data[0], data[1], 64, OTP_CTRL_STATUS_DAI_ERROR_MASK);
```
Do you have to clear the previous error before attempting to write?

moidx commented on src/integration/test_suites/smoke_test_fc_secret_partition_zeroization/smoke_test_fc_secret_partition_zeroization.c:
```diff
@@ -0,0 +1,168 @@
+
+  // Step 8. Read back the digest from the partition's digest register and verify it
+  //         is non-zero.
+  digest[0] = lsu_read_32(OTP_CTRL_SECRET_PROD_PARTITION_0_DIGEST_DIGEST_0);
```
Need to update to read from the correct SOC offset.

moidx commented on src/integration/test_suites/smoke_test_fc_secret_partition_zeroization/smoke_test_fc_secret_partition_zeroization.c:
```diff
@@ -0,0 +1,168 @@
+  // Step 9. Apply zeroization multiple_times
+  for(uint32_t num_times = 0; num_times < zeroize_num_times; num_times++) {
+    // Zeroize fuse.
+    for (uint32_t addr = partitions[VENDOR_SECRET_PROD_PARTITION].address; addr < partitions[VENDOR_SECRET_PROD_PARTITION].digest_address; addr += (partitions[VENDOR_SECRET_PROD_PARTITION].granularity / 8)) {
+      dai_zer(addr, &read_data[0], &read_data[1], partitions[VENDOR_SECRET_PROD_PARTITION].granularity, 0);
+      if (read_data[0] != 0xFFFFFFFF || (partitions[VENDOR_SECRET_PROD_PARTITION].granularity> 32 && read_data[1] != 0xFFFFFFFF)) {
+        VPRINTF(LOW, "ERROR: [partitions[SECRET_LC_TRANSITION_PARTITION] ]fuse is not zeroized\n");
```
This seems to be zeroizing the wrong partition. Also, the code can be simplified as the granularity of the partition is fixed.

moidx commented on src/integration/test_suites/smoke_test_fc_secret_partition_zeroization/smoke_test_fc_secret_partition_zeroization.yml:
```diff
@@ -0,0 +1,9 @@
+---
+seed: ${PLAYBOOK_RANDOM_SEED}
+testname: smoke_test_fc_secret_partition_zeroization
+pre-exec: |
+  echo "Running pre_exec for [smoke_test_fc_secret_partition_zeroization]"
```
Missing: `$CALIPTRA_SS_ROOT/tools/scripts/fuse_ctrl_script/gen_fuse_ctrl_vmem.py` invocation.

andreaskurth commented on src/integration/test_suites/smoke_test_fc_secret_partition_zeroization/smoke_test_fc_secret_partition_zeroization.c:
```diff
@@ -0,0 +1,196 @@
+
+  // Step 8. Read back the digest from the partition's digest register and verify it
+  //         is non-zero.
+  digest[0] = lsu_read_32(SOC_OTP_CTRL_SECRET_PROD_PARTITION_0_DIGEST_DIGEST_0);
+  digest[1] = lsu_read_32(SOC_OTP_CTRL_SECRET_PROD_PARTITION_0_DIGEST_DIGEST_1);
```
VENDOR_SECRET instead

andreaskurth replied:
`SOC_OTP_CTRL_VENDOR_SECRET_PROD_PARTITION_DIGEST_DIGEST_0` and `SOC_OTP_CTRL_VENDOR_SECRET_PROD_PARTITION_DIGEST_DIGEST_1`
